File: kevoree-core/org.kevoree.core/src/main/java/org/kevoree/core/deploy/KevoreeSeqDeployPhase.java
package org.kevoree.core.deploy;

import org.kevoree.api.PrimitiveCommand;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.kevoree.log.Log;
import org.kevoree.core.KevoreeCoreBean;
import org.kevoree.api.telemetry.TelemetryEvent;

public class KevoreeSeqDeployPhase extends KevoreeDeployPhase {
    
    public KevoreeSeqDeployPhase(KevoreeCoreBean core) {
    	super(core);
    }
    
    public boolean execute() {
    	if (primitives.isEmpty()) {
    		return true;
    	} else {
    		PrimitiveCommand lastCmd = null;
    		try {
    			boolean result = true;
    			for (PrimitiveCommand cmd : primitives) {
    				lastCmd = cmd;
    				result = cmd.execute();
    				if (!result) {
        				if (originCore.isAnyTelemetryListener()) {
        					originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR, "Cmd:["+cmd.toString()+"]", null);
        				}
        				Log.error("Error while executing primitive command {}", cmd);
        				break;
    				}
    			}
    			return result;
    		} catch (Throwable e) {
                if (originCore.isAnyTelemetryListener()) {
                    originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR, "Cmd:["+lastCmd.toString()+"]", e);
                }
                Log.error("Exception while executing primitive command {} ", e, lastCmd);
                e.printStackTrace();
                return false;
    		}
    	}
    }

    public void rollback() {
    	Log.trace("Rollback phase");
        if (successor != null) {
            Log.trace("Rollback sucessor first");
            successor.rollback();
        }
        if (!rollbackPerformed) {
        	List<PrimitiveCommand> clonedPrimitives = new ArrayList<PrimitiveCommand>(primitives);
        	Collections.reverse(clonedPrimitives);
            for (PrimitiveCommand c : clonedPrimitives) {
                try {
                    Log.trace("Undo adaptation command {} ", c.getClass().getName());
                    c.undo();
                } catch (Exception e) {
                    if (originCore.isAnyTelemetryListener()) {
                        originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR, "Error while executing primitive command {}", e);
                    }
                    Log.error("Error while executing primitive command {}", c);
                }
            }
            rollbackPerformed = true;
        }	
    }
}

File: kevoree-core/org.kevoree.core/src/main/java/org/kevoree/core/impl/deploy/KevoreeSeqDeployPhase.kt
package org.kevoree.core.impl.deploy

import org.kevoree.api.PrimitiveCommand
import java.util.concurrent.Callable
import java.util.ArrayList
import org.kevoree.log.Log
import java.util.concurrent.TimeUnit
import org.kevoree.core.impl.KevoreeCoreBean
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import org.kevoree.api.telemetry.TelemetryEvent

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 04/12/2013
 * Time: 09:26
 */

class KevoreeSeqDeployPhase(val originCore : KevoreeCoreBean) : KevoreeDeployPhase {
    var primitives: MutableList<PrimitiveCommand> = ArrayList<PrimitiveCommand>()
    var maxTimeout: Long = 30000
    fun setMaxTime(mt: Long) {
        maxTimeout = Math.max(maxTimeout, mt)
    }
    override var sucessor: KevoreeDeployPhase? = null

    override fun populate(cmd: PrimitiveCommand) {
        primitives.add(cmd)
        rollbackPerformed = false
    }

    override fun runPhase(): Boolean {
        if (primitives.size == 0) {
            return true
        }
        var lastPrimitive : PrimitiveCommand? = null
        try {
            var result = true

            for(primitive in primitives) {
                lastPrimitive = primitive
                result = primitive.execute()
                if(!result){
                    if(originCore.isAnyTelemetryListener()){
                        originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR,"Cmd:["+primitive.toString()+"]",null)
                    }
                    //originCore.broadcastTelemetry("warn","Error during execution of "+primitive, e.toString())
                    Log.warn("Error during execution of {}",primitive)
                    break;
                }
            }
            return result
        } catch (e:Throwable){
            if(originCore.isAnyTelemetryListener()){
                try {
                    originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR,"Cmd:["+lastPrimitive.toString()+"]",e)
                } catch (e: Throwable){
                   e.printStackTrace()
                }
            }
            e.printStackTrace()
            return false
        }
    }

    var rollbackPerformed = false

    override fun rollBack() {
        Log.trace("Rollback phase")
        if (sucessor != null) {
            Log.trace("Rollback sucessor first")
            sucessor?.rollBack()
        }
        if(!rollbackPerformed){
            // SEQUENCIAL ROOLBACK
            for(c in primitives.reverse()){
                try {
                    Log.trace("Undo adaptation command {} ", c.javaClass.getName())
                    c.undo()
                } catch (e: Exception) {
                    originCore.broadcastTelemetry(TelemetryEvent.Type.LOG_ERROR,"Exception during rollback", e)
                    //Log.warn("Exception during rollback", e)
                }
            }
            rollbackPerformed = true
        }
    }
}